documents/model.go

```diff
@@ -740,17 +749,197 @@ func (m *CoreDocumentModel) ValidateDocumentAccess(docReq *p2ppb.GetDocumentRequ
 }
 
 // isNFTInRole checks if the given nft(registry + token) is part of the core document role.
-func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []byte) bool {
+// If found, returns the index of the nft in the role and true
+func isNFTInRole(role *coredocumentpb.Role, registry common.Address, tokenID []byte) (nftIdx int, found bool) {
 	enft, err := ConstructNFT(registry, tokenID)
 	if err != nil {
-		return false
+		return nftIdx, false
 	}
 
-	for _, n := range role.Nfts {
+	for i, n := range role.Nfts {
 		if bytes.Equal(n, enft) {
-			return true
+			return i, true
 		}
 	}
 
-	return false
+	return nftIdx, false
+}
+
+// AddNFT returns a new CoreDocument model with nft added to the Core document. If grantReadAccess is true, the nft is added
+// to the read rules.
+func (m *CoreDocumentModel) AddNFT(grantReadAccess bool, registry common.Address, tokenID []byte) (*CoreDocumentModel, error) {
+	data := m.Document.EmbeddedData
+	ndm, err := m.prepareNewVersion(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ndm.Document.EmbeddedData = data
+	cd := ndm.Document
+	nft := getStoredNFT(cd.Nfts, registry.Bytes())
+	if nft == nil {
+		nft = new(coredocumentpb.NFT)
+		// add 12 empty bytes
+		eb := make([]byte, 12, 12)
+		nft.RegistryId = append(registry.Bytes(), eb...)
+		cd.Nfts = append(cd.Nfts, nft)
+	}
+	nft.TokenId = tokenID
+
+	if grantReadAccess {
+		err = ndm.AddNFTToReadRules(registry, tokenID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ndm, ndm.setCoreDocumentSalts()
+}
+
+// IsNFTMinted checks if the there is an NFT that is minted against this document in the given registry.
+func (m *CoreDocumentModel) IsNFTMinted(tokenRegistry TokenRegistry, registry common.Address) bool {
```

Comment:
I think this function must not exist in a model
A model should not have any runtime dependancies, it makes them hard to test. We can make this part of some nft utility and call it when needed from outside the model.

Reply:
I agree with you but checking the NFT is minted requires us to expose the `coredocumentpb.CoreDocument` which is what we tried avoiding with coredocmodel refactor.

Reply:
Wouldn't we just mock the tokenRegistry here? I think this is fine the way you did it Ved.

Reply:
@lucasvo The function looks like it was hacked to make it testable, its a code smell for me :) But if it's hard to change its fine. Alternative is to define a method to get the NFT ID check externally.

Reply:
I know, this is something we discussed already in the discussion for coredoc refactor as well, a model having